[Core][ModelPart] Improve Prints when distributed

philbucher opened the pull request:
This PR adds prints of the `Communicator` in the ModelPart in case it is distributed:
~~~
-default- model part
    Buffer Size : 1
    Number of tables : 0
    Number of sub model parts : 0
    Distributed over 3 processes
    Current solution step index : 0
    DOMAIN_SIZE : 3

    Mesh 0 :
        Number of Nodes       : 3
        Number of Properties  : 0
        Number of Elements    : 0
        Number of Conditions  : 0
        Number of Constraints : 0

    Local Mesh  : 
        Number of Nodes       : 3
        Number of Properties  : 0
        Number of Elements    : 0
        Number of Conditions  : 0
        Number of Constraints : 0
    Ghost Mesh  : 
        Number of Nodes       : 0
        Number of Properties  : 0
        Number of Elements    : 0
        Number of Conditions  : 0
        Number of Constraints : 0
    Interface Mesh  : 
        Number of Nodes       : 0
        Number of Properties  : 0
        Number of Elements    : 0
        Number of Conditions  : 0
        Number of Constraints : 0
~~~

We can discuss what the communicator should print, since it might be quite verbose

philbucher committed:
deleted duplicated method from mpi-communicator

philbucher committed:
improving prints in Communicator

philbucher committed:
printing info in case of distributed MP

pooyan-dadvand commented:
@KratosMultiphysics/technical-committee delegate full decision to @KratosMultiphysics/implementation-committee

philbucher committed:
Merge remote-tracking branch 'origin/master' into core/mp-prints-for-distributed

adityaghantasala commented:
IMO the communicator can also print the number of neighbors too. This is useful sometimes.

philbucher commented:
I don’t think the communicator has neighbors
Or what do you mean?

adityaghantasala commented:
It has no ? 

https://github.com/KratosMultiphysics/Kratos/blob/ebab4ff7b66ec1d83515fe9ca7aa5860847bd368/kratos/includes/communicator.h#L470

philbucher commented:
Ah that
I will check
No clue what it is :)

philbucher commented:
I checked, the neighbors are the colors
IMO they should not be printed by default, I don't think this adds valuable information except for when debugging
In a large case this can be many, but what would you use this information for?

Note that already now the `Communicator` prints many things

philbucher committed:
Merge remote-tracking branch 'origin/master' into core/mp-prints-for-distributed

adityaghantasala commented:
I mean you need not print the colors themselves, but just how many is sufficient and informative ! 
Also CD/CI is failing .. can you check !

philbucher commented:
> I mean you need not print the colors themselves, but just how many is sufficient and informative !

ah ok that I can do

> Also CD/CI is failing .. can you check !

I know, I wrote tests for the printing some time ago
I will fix them once the @KratosMultiphysics/implementation-committee has reached a decision, otherwise I have to do it multiple times

ping @KratosMultiphysics/implementation-committee

adityaghantasala commented:
Just a small comment 

> Distributed over 3 processes

in the above, is the total number of processors correct ? 

Here one could print : 

> Total number of processors : 3
> Current modelpart is distributed on : 2 (number of colors)

philbucher commented on kratos/sources/model_part.cpp:
```diff
@@ -1786,6 +1786,9 @@ void ModelPart::PrintData(std::ostream& rOStream) const
     rOStream << "    Number of sub model parts : " << NumberOfSubModelParts() << std::endl;
     if (!IsSubModelPart()) {
+        if (IsDistributed()) {
+        rOStream << "    Distributed over " << mpCommunicator->TotalProcesses() << " processes" << std::endl;
```
@adityaghantasala this is the size of the `MPI_Comm` with which the data-communicator was created
By default this is the size of `MPI_COMM_WORLD`, but with a sub-communicator it would be the size of this sub-communicator

adityaghantasala replied:
Ya exactly.

GuillermoCasas commented:
On behalf of the @KratosMultiphysics/implementation-committee, we think the suggested change by @adityaghantasala would complete this design, making it clearer for the user. That is, we suggest to replace 'Distributed over n processes', where n is the total number of processes in the particualr communicator by the following two messages:

- 'Number of processes (colors) over which ModelPart is distributed: k'
- 'Total number of processes: n'

rubenzorrilla commented:
> On behalf of the @KratosMultiphysics/implementation-committee, we think the suggested change by @adityaghantasala would complete this design, making it clearer for the user. That is, we suggest to replace 'Distributed over n processes', where n is the total number of processes in the particualr communicator by the following two messages:
> 
>     * 'Number of processes (colors) over which ModelPart is distributed: k'
> 
>     * 'Total number of processes: n'

Just to say that I found that this could be too much verbose, specially in long model part names case (I assume that `ModelPart` is the model part name). I'd be in favour of @adityaghantasala 's first suggestion.

philbucher committed:
Merge remote-tracking branch 'origin/master' into core/mp-prints-for-distributed

philbucher committed:
remove communicator print
Too verbose, and only really needed for debugging, for which it can be manually printed

philbucher commented on kratos/mpi/includes/mpi_communicator.h:
```diff
@@ -1188,29 +1188,6 @@ template<OperationType TOperation> struct Operation
 
-    /// Print information about this object.
-
-    void PrintInfo(std::ostream& rOStream) const override
```
I am removing this since it is a duplicate from the baseclass

philbucher commented on kratos/mpi/includes/mpi_communicator.h:
```diff
@@ -1188,29 +1188,6 @@ template<OperationType TOperation> struct Operation
-
-    /// Print object's data.
-
-    void PrintData(std::ostream& rOStream) const override
```
same

philbucher reviewed (commented):
> On behalf of the @KratosMultiphysics/implementation-committee, we think the suggested change by @adityaghantasala would complete this design, making it clearer for the user. That is, we suggest to replace 'Distributed over n processes', where n is the total number of processes in the particualr communicator by the following two messages:
> 
>     * 'Number of processes (colors) over which ModelPart is distributed: k'
> 
>     * 'Total number of processes: n'

@KratosMultiphysics/implementation-committee 
I gave this another go and came to the following conclusion:
- Printing the data of the `Communicator` is too verbose to always do it, I hence removed it again. It is mostly needed for debugging anyway, and for that purpose it can be printed manually
- I think you misunderstood sth related to the colors: The size of the Communicator IS the number of processes over which the modelpart is distributed! Remember that this can be a Subcomm of `MPI_COMM_WORLD`.
- The total number of processes is printed at the beginning of the Simulation under the Kratos-logo, I don't think it makes sense to reprint it

adityaghantasala commented:
> The total number of processes is printed at the beginning of the Simulation under the Kratos-logo, I don't think it makes sense to reprint it

I am still in favour of printing the total number of processors .. this because one can print the comm info during the simulation at some point. And there the LOGO is not visible with all the previously printed output .. it will be convenient to see the total number of processes ...  after all we are printing the communicator information.

philbucher commented:
> > The total number of processes is printed at the beginning of the Simulation under the Kratos-logo, I don't think it makes sense to reprint it
> 
> I am still in favour of printing the total number of processors .. this because one can print the comm info during the simulation at some point. And there the LOGO is not visible with all the previously printed output .. it will be convenient to see the total number of processes ... after all we are printing the communicator information.

ok that makes sense...
Will add it

philbucher committed:
adding print for total processes

philbucher commented on kratos/sources/model_part.cpp:
```diff
@@ -1786,6 +1787,10 @@ void ModelPart::PrintData(std::ostream& rOStream) const
     if (!IsSubModelPart()) {
+        if (IsDistributed()) {
+            rOStream << "    Distributed over " << mpCommunicator->TotalProcesses() << " processes" << std::endl;
+            rOStream << "    Total number of processes: " << ParallelEnvironment::GetDataCommunicator("World").Size() << std::endl;
```
@adityaghantasala I added it, this is the same as in the kernel:
https://github.com/KratosMultiphysics/Kratos/blob/457a15824bd40485fe612c69a56559a058dbdc10/kratos/sources/kernel.cpp#L143-L144

however this means that we have to include another header: `#include "includes/parallel_environment.h"`. I guess this is ok since it is a cpp but still...

**Alternative:**
this would be the alternative without additional includes:
```suggestion
            rOStream << "    Total number of processes: " << DataCommunicator::GetDefault().Size() << std::endl;
```

however someone could change the default `DataCommunicator` then this would be wrong

**Conclusion:**
I am not sure if we should add it. After all the number of processes is sth critical for an MPI-run, one does not easily "forget" that...
What do you think?

adityaghantasala commented:
@philbucher  I think we miss understood each other ... I had the following implementation in mind 

```cpp
if (IsDistributed()) {
  rOStream << "    Total number of processes: " << mpCommunicator->TotalProcesses()  << std::endl;
  rOStream << "    Distributed over " << mpCommunicator->GetNumberOfColors() << " processes" << std::endl;
}
```

philbucher committed:
update after speaking

philbucher commented:
@adityaghantasala I updated it after we spoke, can you check please?

philbucher committed:
missing space

adityaghantasala commented:
@philbucher  Looks good ... ! I am approving but wait till tomorrow evening as @KratosMultiphysics/implementation-committee  we have a short discussion tomorrow and confirm this !

philbucher commented:
did you have the meeting?

pablobecker commented:
On behalf of the @KratosMultiphysics/implementation-committee we reached consensus on this matter. We think this PR is ready to be merged.

philbucher commented:
thx for the reviews!
